revocation/revocation.go

```diff
@@ -101,49 +110,142 @@ func NewWithOptions(opts Options) (Validator, error) {
 	}
 
 	return &revocation{
-		httpClient:       opts.OCSPHTTPClient,
+		ocspHTTPClient:   opts.OCSPHTTPClient,
+		crlHTTPClient:    opts.CRLHTTPClient,
 		certChainPurpose: opts.CertChainPurpose,
 	}, nil
 }
 
 // Validate checks the revocation status for a certificate chain using OCSP and
-// returns an array of CertRevocationResults that contain the results and any
-// errors that are encountered during the process
+// CRL if OCSP is not available. It returns an array of CertRevocationResults
+// that contain the results and any errors that are encountered during the
+// process.
 //
-// TODO: add CRL support
-// https://github.com/notaryproject/notation-core-go/issues/125
+// This function tries OCSP and falls back to CRL when:
+// - OCSP is not supported by the certificate
+// - OCSP returns an unknown status
+//
+// NOTE: The certificate chain is expected to be in the order of leaf to root.
 func (r *revocation) Validate(certChain []*x509.Certificate, signingTime time.Time) ([]*result.CertRevocationResult, error) {
 	return r.ValidateContext(context.Background(), ValidateContextOptions{
 		CertChain:            certChain,
 		AuthenticSigningTime: signingTime,
 	})
 }
 
-// ValidateContext checks the revocation status for a certificate chain using
-// OCSP and returns an array of CertRevocationResults that contain the results
-// and any errors that are encountered during the process
+// ValidateContext checks the revocation status for a certificate chain using OCSP and
+// CRL if OCSP is not available. It returns an array of CertRevocationResults
+// that contain the results and any errors that are encountered during the
+// process.
+//
+// This function tries OCSP and falls back to CRL when:
+// - OCSP is not supported by the certificate
+// - OCSP returns an unknown status
 //
-// TODO: add CRL support
-// https://github.com/notaryproject/notation-core-go/issues/125
+// NOTE: The certificate chain is expected to be in the order of leaf to root.
 func (r *revocation) ValidateContext(ctx context.Context, validateContextOpts ValidateContextOptions) ([]*result.CertRevocationResult, error) {
 	if len(validateContextOpts.CertChain) == 0 {
 		return nil, result.InvalidChainError{Err: errors.New("chain does not contain any certificates")}
 	}
+	certChain := validateContextOpts.CertChain
 
-	return ocsp.CheckStatus(ocsp.Options{
-		CertChain:        validateContextOpts.CertChain,
-		CertChainPurpose: r.certChainPurpose,
-		SigningTime:      validateContextOpts.AuthenticSigningTime,
-		HTTPClient:       r.httpClient,
-	})
+	if err := x509util.ValidateChain(certChain, r.certChainPurpose); err != nil {
+		return nil, err
+	}
+
+	ocspOpts := ocsp.CertCheckStatusOptions{
+		HTTPClient:  r.ocspHTTPClient,
+		SigningTime: validateContextOpts.AuthenticSigningTime,
+	}
+
+	crlOpts := crl.CertCheckStatusOptions{
+		HTTPClient:  r.crlHTTPClient,
+		SigningTime: validateContextOpts.AuthenticSigningTime,
+	}
+
+	panicChan := make(chan any, len(certChain))
```

Comment:
nit: will it be better to have the following pattern?
```suggestion
	panicChan := make(chan any, len(certChain))
	defer close(panicChan)
```

Reply:
Updated.